src/main/java/org/springframework/data/mapping/model/ReflectionEntityInstantiator.java

```diff
@@ -44,39 +46,58 @@
 	public <T, E extends PersistentEntity<? extends T, P>, P extends PersistentProperty<P>> T createInstance(E entity,
 			ParameterValueProvider<P> provider) {
 
-		var constructor = entity.getPersistenceConstructor();
+		var creator = entity.getEntityCreator();
 
-		if (constructor == null) {
-
-			try {
-				Class<?> clazz = entity.getType();
-				if (clazz.isArray()) {
-					var ctype = clazz;
-					var dims = 0;
-					while (ctype.isArray()) {
-						ctype = ctype.getComponentType();
-						dims++;
-					}
-					return (T) Array.newInstance(clazz, dims);
-				} else {
-					return BeanUtils.instantiateClass(entity.getType());
-				}
-			} catch (BeanInstantiationException e) {
-				throw new MappingInstantiationException(entity, Collections.emptyList(), e);
-			}
+		if (creator == null) {
+			return instantiateClass(entity);
 		}
-		var parameterCount = constructor.getConstructor().getParameterCount();
+		var parameterCount = creator.getParameterCount();
 
 		var params = parameterCount == 0 ? EMPTY_ARGS : new Object[parameterCount];
 		var i = 0;
-		for (Parameter<?, P> parameter : constructor.getParameters()) {
+		for (Parameter<?, P> parameter : creator.getParameters()) {
 			params[i++] = provider.getParameterValue(parameter);
 		}
 
+		if (creator instanceof FactoryMethod method) {
```

Comment:
Again, I think this should be on the `EntityCreator` interface to avoid the instance of checks and pulling out implementation details.